File: bst/bst.cpp
/*
  Rakesh Kumar
  13/06/2021
  Never stop learning
 */

#include <bits/stdc++.h>

struct Node {
    Node() {}
    Node(int val) : val(val) {}

    int val = 0;
    Node* left = nullptr;
    Node* right = nullptr;
};

class BST {
public:
    Node* insert(Node* root, int val) {
        if (root == nullptr)
            root = new Node(val);
        else {
            Node* node = root;
            while (node) {
                if (val < node->val) {
                    if (node->left == nullptr) {
                        node->left = new Node(val);
                        break;
                    } else
                        node = node->left;
                } else {
                    if (node->right == nullptr) {
                        node->right = new Node(val);
                        break;
                    } else
                        node = node->right;
                }
            }
        }
        return root;
    }

    int height(Node* root) {
        if (root == nullptr)
            return 0;
        int l = height(root->left) + 1;
        int r = height(root->right) + 1;
        return std::max(l, r);
    }

    void bfs(Node* root) {
        printf("--------------------------------------------------\n");
        std::queue<Node*> q;
        q.push(root);
        while (!q.empty()) {
            for (int i = static_cast<int>(q.size()); i > 0; --i) {
                root = q.front(); q.pop();
                printf("%d ", root->val);
                if (root->left) q.push(root->left);
                if (root->right) q.push(root->right);
            }
            printf("\n");
        }
        printf("--------------------------------------------------\n");
    }
};

Node* build_bst() {
    BST bst;
    Node* root = nullptr;
    root = bst.insert(root, 100);
    root = bst.insert(root, 50);
    root = bst.insert(root, 150);
    root = bst.insert(root, 25);
    root = bst.insert(root, 175);
    root = bst.insert(root, 75);
    root = bst.insert(root, 125);
    root = bst.insert(root, 20);
    root = bst.insert(root, 200);
    root = bst.insert(root, 10);
    root = bst.insert(root, 5);
    return root;
}

void test_insert() {
    BST bst;
    Node* root = build_bst();
    bst.bfs(root);
}

void test_height() {
    BST bst;
    Node* root = build_bst();
    printf("height == [%d]\n", bst.height(root));
}

int main() {
    test_insert();
    test_height();

    return 0;
}

File: bst/bst.py
'''
    Never stops learning
    Rakesh Kumar
    23/06/2021
'''

import common

class BST(object):
    def insert_iter(self, root, val):
        if root == None:
            root = common.Node(val)
        else:
            node = root
            while node:
                if val < node.val:
                    if node.left == None:
                        node.left = common.Node(val)
                        break
                    else:
                        node = node.left
                else:
                    if node.right == None:
                        node.right = common.Node(val)
                        break
                    else:
                        node = node.right
        return root

    def insert_rec(self, root, val):
        if root == None:
            root = common.Node(val)
        else:
            if val < root.val:
                root.left = self.insert_rec(root.left, val)
            else:
                root.right = self.insert_rec(root.right, val)
        return root

    def preorder(self, root):
        if root == None:
            return
        print(root.val, end=' ')
        self.preorder(root.left)
        self.preorder(root.right)

    def inorder(self, root):
        if root == None:
            return
        self.inorder(root.left)
        print(root.val, end=' ')
        self.inorder(root.right)

    def postorder(self, root):
        if root == None:
            return
        self.postorder(root.left)
        self.postorder(root.right)
        print(root.val, end=' ')

    def preorder_data(self, root):
        data = []
        self._preorder_data(root, data)
        return data

    def inorder_data(self, root):
        data = []
        self._inorder_data(root, data)
        return data

    def postorder_data(self, root):
        data = []
        self._postorder_data(root, data)
        return data

    def bfs(self, root):
        print('--------------------------------------------------------')
        if root is not None:
            q = []
            q.append(root)
            while len(q) > 0:
                i = len(q)
                while i > 0:
                    root = q.pop(0)
                    print(root.val, end = ' ')
                    if root.left is not None:
                        q.append(root.left)
                    if root.right is not None:
                        q.append(root.right)
                    i -= 1
                print('')
        print('--------------------------------------------------------')

    def _preorder_data(self, root, data):
        if root == None:
            return
        data.append(root.val)
        self._preorder_data(root.left, data)
        self._preorder_data(root.right, data)

    def _inorder_data(self, root, data):
        if root == None:
            return
        self._inorder_data(root.left, data)
        data.append(root.val)
        self._inorder_data(root.right, data)

    def _postorder_data(self, root, data):
        if root == None:
            return
        self._postorder_data(root.left, data)
        self._postorder_data(root.right, data)
        data.append(root.val)

